public_html/wp-content/plugins/wcpt/wcpt-event/notification.php

```diff
@@ -0,0 +1,141 @@
+	define( 'COMMUNITY_TEAM_SLACK', '@vedjain' );
+	define( 'COMMUNITY_EVENTS_SLACK', '@vedjain' );
+}
+
+/**
+ * Send attachment to Make WordPress slack. Will be used to send event notifications to community channels
+ *
+ * @param string $channel Name of the channel we want to send the notification to.
+ * @param array  $attachment Attachment object
+ *
+ * @return bool|string
+ */
+function wcpt_slack_notify ( $channel, $attachment ) {
+	if ( ! class_exists( 'Dotorg\Slack\Send' ) ) {
+		return false;
+	}
+
+	if ( ! defined( 'SLACK_ERROR_REPORT_URL' ) ) {
+		return false;
+	}
+
+	$send = new \Dotorg\Slack\Send( SLACK_ERROR_REPORT_URL );
+	$send->add_attachment( $attachment );
+	return $send->send( $channel );
+}
+
+/**
+ * Send attachment in #community-teams channel
+ *
+ * @param array $attachment An object containing slack attachment
+ *
+ * @return bool|string
+ */
+function notify_in_team_channel ( $attachment ) {
+	return wcpt_slack_notify( COMMUNITY_TEAM_SLACK, $attachment );
+}
+
+/**
+ * Send attachment in #community-events channel
+ *
+ * @param array $attachment An object containing slack attachment
+ *
+ * @return bool|string
+ */
+function notify_in_events_channel ( $attachment ) {
+	return wcpt_slack_notify( COMMUNITY_EVENTS_SLACK, $attachment );
+}
+
+/**
+ * Creates attachment that can be sent using slack notification.
+ * See the structure of attachment here: https://api.slack.com/docs/message-attachments
+ *
+ * @param string $message Main text to send in the notification
+ * @param string $event_label Label for the event. Would probably be one of `WordCamp` or `Meetup`.
+ *
+ * @return array
+ */
+function create_event_attachment ( $message, $title ) {
+	// TODO: Change this based on community suggestion
+	// Not translating because this will be send to Slack.
+	return array(
+		"title" => $title,
+		"text" => $message,
+	);
+}
+
+function create_event_status_attachment( $message, $event_id, $title ) {
+	// TODO: Change this based on community suggestion
+	$props = get_props_for_event( $event_id );
+
+	// remove bot user `wordcamp`
+	$props = array_diff( $props, array( 'wordcamp' ) );
+
+	$props_string = implode( ", ", $props );
+	return array(
+		"title" => $title,
+		"text" => $message,
+		"fields" => array(
+			array(
+				"title" => "Props",
+				"value" => $props_string,
+			),
+		),
+	);
+}
+
+/**
+ * Get and array of usernames for everyone contributed to vetting an event application.
+ * This includes people who have changed statuses, added notes etc.
+ * Currently supports WordCamp and Meetups
+ *
+ * @param int $event_id Id of the event to fetch props for.
+ *
+ * @return array Array of usernames of people who have participated in vetting this application
+ */
+function get_props_for_event ( $event_id ) {
+	$user_ids = array();
+
+	$status_change_logs = get_post_meta( $event_id, '_status_change' ) ?? array();
+
+	$user_ids = array_merge( $user_ids, wp_list_pluck( $status_change_logs, 'user_id' ) );
```

Comment:
so I am initializing with empty array above `?? array();` in case `_status_change` log is not present. wp_list_pluck should be fine with empty array I think? is that not the case

Reply:
Here's the basic logic in `wp_list_pluck`:

```php
foreach ( $this->output as $key => $value ) {
	if ( is_object( $value ) ) {
		$this->output[ $key ] = $value->$field;
	} else {
		$this->output[ $key ] = $value[ $field ];
	}
}
```

So it looks like if `$value[ $field ]` doesn't exist, it would throw the `undefined index` warning. But I haven't tested it.

Reply:
oh yes, if the field is not present there would be notice. but given that status log is created by our code, it will always have those fields. keeping that in mind there should be a notice if the field is not present.
otherwise i cannot use `wp_list_pluck` and have to use a loop to get all the elements.

Reply:
> given that status log is created by our code, it will always have those fields

Right, but if there aren't any status log entries yet, so the default empty array is returned via `?? array();`, would that throw a notice? I guess maybe the foreach loop wouldn't even get far enough to detect an undefined index if the array has no items...

Reply:
looks like default empty array will not throw a notice. and you are correct that foreach loop wouldn't event get far enough to detect an undefined index if the array has not items.